infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/ParallelDockerImagesPreparer.java

```diff
@@ -0,0 +1,323 @@
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.eclipse.che.infrastructure.docker.client.DockerConnector;
+import org.eclipse.che.infrastructure.docker.client.DockerFileException;
+import org.eclipse.che.infrastructure.docker.client.ProgressMonitor;
+import org.eclipse.che.infrastructure.docker.client.UserSpecificDockerRegistryCredentialsProvider;
+import org.eclipse.che.infrastructure.docker.client.exception.ImageNotFoundException;
+import org.eclipse.che.infrastructure.docker.client.json.Filters;
+import org.eclipse.che.infrastructure.docker.client.params.BuildImageParams;
+import org.eclipse.che.infrastructure.docker.client.params.ListImagesParams;
+import org.eclipse.che.infrastructure.docker.client.params.PullParams;
+import org.eclipse.che.infrastructure.docker.client.params.TagParams;
+import org.eclipse.che.infrastructure.docker.client.parser.DockerImageIdentifier;
+import org.eclipse.che.infrastructure.docker.client.parser.DockerImageIdentifierParser;
+import org.eclipse.che.workspace.infrastructure.docker.exception.SourceNotFoundException;
+import org.eclipse.che.workspace.infrastructure.docker.logs.MachineLoggersFactory;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class allows to make parallel prepare (build or download) of docker images for workspace
+ * being started.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class ParallelDockerImagesPreparer {
+
+  private static final Logger LOG = getLogger(ParallelDockerImagesPreparer.class);
+  private static final String PARALLEL_NUMBER_PROPERTY = "che.infra.docker.max_pull_thread";
+
+  private final RuntimeIdentity identity;
+  private final MachineLoggersFactory machineLoggersFactory;
+  private final boolean doForcePullImage;
+  private final UserSpecificDockerRegistryCredentialsProvider dockerCredentials;
+  private final DockerConnector dockerConnector;
+  private final ThreadPoolExecutor executor;
+
+  @Inject
+  public ParallelDockerImagesPreparer(
+      @Assisted RuntimeIdentity identity,
+      @Named("che.docker.always_pull_image") boolean doForcePullImage,
+      @Named(PARALLEL_NUMBER_PROPERTY) int parallelPullsNumber,
+      UserSpecificDockerRegistryCredentialsProvider dockerCredentials,
+      DockerConnector dockerConnector,
+      MachineLoggersFactory machineLoggersFactory) {
+    this.identity = identity;
+    this.doForcePullImage = doForcePullImage;
+    this.dockerCredentials = dockerCredentials;
+    this.dockerConnector = dockerConnector;
+    this.machineLoggersFactory = machineLoggersFactory;
+
+    ThreadFactory factory =
+        new ThreadFactoryBuilder()
+            .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+            .setNameFormat(getClass().getSimpleName())
```

Comment:
Use `%d` to show the number of thread otherwise all the threads would look the same and analysis of logs will be complicated.